## Lab 03.01 - Introduction to Pandas and Data Operations
In this lab, you'll learn how to use the pandas library to analyze real-world sales data. By the end of this lab, you'll be comfortable with the fundamental operations in pandas.

### Part 0 - Intro to Pandas
Before we dive into using pandas, let's understand what it is and why it's important in data science.

To answer the questions, edit the markdown cell and put your answer below the question. 

**Make sure to save the markdown cell, by pressing the ✓ (check) icon in the top right after answering the questions**

Research and critically think about the following questions:

##### Question 00
What is Pandas, and what are its primary functions in data visualization for data science?
- **Answer:** Pandas is a python module that allows the analysis, sorting, and filtering of data sets

##### Question 01
What does "data manipulation" mean in the context of pandas and data science field?
- **Answer:** In the context of data science, data manipulation is taking raw data and converting it for a specific purpose, like visualizing it or isolating specific desired data.

##### Question 02
What kind of files can pandas read? List at least 3:
- **Answer:** pandas can read csv, excel, and json files.

#### 0.0 - Hands-On Exploration
Let's start by importing pandas and exploring its basic functionality. Type these commands and write down what you observe:

First, let's install the matplotlib library, in your terminal run the following command:

```bash
pip3 install pandas
```

Once installed, let's start by importing pandas and exploring its basic functionality. 

In [2]:
import pandas as pd

Now lets run create a list, and see what happens when we use it with pandas.

In [3]:
# Create a simple list of numbers
numbers = [1, 2, 3, 4, 5]

# Convert it to a pandas Series
series = pd.Series(numbers)

# Print the result
print(series)

0    1
1    2
2    3
3    4
4    5
dtype: int64


##### Question 03
What did you notice about the output format?
- **Answer:** The list items are printed on 
new lines.
##### Question 04
How is this different from a regular Python list? What are the numbers on the left side?
- **Answer:** The numbers on the left represent the index of the list

### Part 1 - Reading and Exploring Data

#### 01.00 - Loading Data
Let's start by loading our sales data that you downloaded alongside this lab. In pandas, we can read various file formats, but CSV (Comma-Separated Values) is one of the most common.

In [4]:
sales_df = pd.read_csv('sales_data.csv')

When using pandas to load data, in ingests it into what is known as a "dataframe", this allows us to do advanced manipulation.

#### 01.01 - Viewing Data Samples
Let's explore the fundamental methods for understanding our data structure and content. We'll look at each method individually and understand what it tells us about our data.

In [9]:
print(sales_df.head())

         Date    Product         Category    Price  Units    Total  Discount  \
0  2023-01-01   Notebook  Office Supplies    18.33      2    35.56      0.03   
1  2023-01-01     Laptop      Electronics  1943.35      4  7773.40      0.00   
2  2023-01-01     Tablet      Electronics   436.05      2   872.10      0.00   
3  2023-01-01  USB Drive      Accessories   116.48      3   349.44      0.00   
4  2023-01-01    Pen Set  Office Supplies    42.14      4   168.56      0.00   

  Region      Sales_Rep Payment_Method Customer_Segment  
0  South   Sarah Wilson     Debit Card   Small Business  
1  North     John Smith     Debit Card   Small Business  
2   East    David Brown  Bank Transfer       Government  
3   East  Lisa Anderson         PayPal       Individual  
4   East    David Brown     Debit Card        Corporate  


In [6]:
print(sales_df.tail())

            Date       Product         Category    Price  Units    Total  \
3766  2023-12-31    Smartwatch      Electronics  1730.72      2  3461.44   
3767  2023-12-31          Desk        Furniture   462.57      1   462.57   
3768  2023-12-31        Laptop      Electronics   560.79      2  1121.58   
3769  2023-12-31  Office Chair        Furniture   116.04      2   232.08   
3770  2023-12-31      Notebook  Office Supplies    18.15      4    72.60   

      Discount Region        Sales_Rep Payment_Method Customer_Segment  
3766       0.0  South  Michael Johnson     Debit Card       Individual  
3767       0.0  North       John Smith     Debit Card       Government  
3768       0.0   East      David Brown     Debit Card   Small Business  
3769       0.0  North       Emma Davis    Credit Card        Corporate  
3770       0.0  North       Emma Davis         PayPal       Individual  


##### Question 04
What is the difference between the `tail()` and `head()` commands? What would be use of either command be?
- **Answer:** The tail function displayed the last items in the csv file while the head function displayed the first items

##### Question 05
What happens when you put a number in the `head()` function? What changed?
- **Answer:** Putting a number prints that index in the csv file 

#### 01.02 - Understanding DataFrame Structure
Let's examine the basic properties of our DataFrame:

In [10]:
print(sales_df.columns)

Index(['Date', 'Product', 'Category', 'Price', 'Units', 'Total', 'Discount',
       'Region', 'Sales_Rep', 'Payment_Method', 'Customer_Segment'],
      dtype='object')


In [11]:
print(sales_df.dtypes)

Date                 object
Product              object
Category             object
Price               float64
Units                 int64
Total               float64
Discount            float64
Region               object
Sales_Rep            object
Payment_Method       object
Customer_Segment     object
dtype: object


##### Question 06
Compare and contrast the outputs of the `columns` and `dtypes` properties. What is similar what is different?
- **Answer:** Both display the catagories of the data set, but the dtypes property displays the data type of the columns as well

##### Question 07
What data type is the 'Date' column? Is this what you expected? 
- **Answer:** The date is an object type, I expected it to be a number

##### Question 08
Why might pandas choose different data types for different columns?
- **Answer:** Depending what is being stored in the columns, pandas chooses a data type that corresponds to it.

#### 01.03 - Data Information Summary
The `info()` method provides a concise summary of our DataFrame:

In [12]:
print(sales_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3771 entries, 0 to 3770
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date              3771 non-null   object 
 1   Product           3771 non-null   object 
 2   Category          3771 non-null   object 
 3   Price             3771 non-null   float64
 4   Units             3771 non-null   int64  
 5   Total             3771 non-null   float64
 6   Discount          3771 non-null   float64
 7   Region            3771 non-null   object 
 8   Sales_Rep         3693 non-null   object 
 9   Payment_Method    3739 non-null   object 
 10  Customer_Segment  3771 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 324.2+ KB
None


##### Question 09
What information does `info()` tell us, how could this information be useful?
- **Answer:** The info functions tells us the range of the data set, the data types used in it, how many columns there are, and how many of each column is present in the set. This information can be useful when looking to isolate data, (eg. looking for only the float values in a data set )

##### Question 10
How much memory is our DataFrame using? 
- **Answer:** Over 324+ KB 

##### Question 11
What does "null" mean? What impact could having "null" values in a dataset?
- **Answer:** null means that there is no data in that columns

#### 01.04 - Numerical Summaries
The `describe()` method provides statistical summaries for numerical columns:

In [13]:
print(sales_df.describe())

             Price        Units         Total     Discount
count  3771.000000  3771.000000   3771.000000  3771.000000
mean    397.093911     2.985150   1159.701854     0.021594
std     514.524434     1.410003   1750.424109     0.041496
min       4.810000     1.000000      4.990000     0.000000
25%      40.165000     2.000000    110.730000     0.000000
50%     134.860000     3.000000    372.870000     0.000000
75%     516.490000     4.000000   1472.525000     0.020000
max    2073.680000     5.000000  10151.100000     0.150000


##### Question 12
What statistics does the `describe()` function give us?
- **Answer:** The describe function gives us the count, mean, standard deviation, minimum and maximum value, and the 25th, 50th, and 75th percentiles of the data.

##### Question 13
Which columns did the `describe()` function perform statistics on? Why? (Hint: why didnt we see 'Product' in the output?)
- **Answer:** The describe function only analyses numerical values (You can't find the mean of a string)

In [14]:
print(sales_df[['Price', 'Units']].describe())

             Price        Units
count  3771.000000  3771.000000
mean    397.093911     2.985150
std     514.524434     1.410003
min       4.810000     1.000000
25%      40.165000     2.000000
50%     134.860000     3.000000
75%     516.490000     4.000000
max    2073.680000     5.000000


##### Question 14
What is different in the codeblock above compared to the previous one. How did this affect the output?
- **Answer:** It only provided the describe details for the columns that were specified.

In [25]:
print(sales_df['Category'].unique())

['Office Supplies' 'Electronics' 'Accessories' 'Furniture']


##### Question 15
How many unique product categories are in our data set?
- **Answer:** There are 4 unique product catagories

In [16]:
print("Output 1:")
print(sales_df['Category'].value_counts())

print('')
print('')

print("Output 2:")
print(sales_df['Category'].value_counts(normalize=True))

Output 1:
Category
Office Supplies    953
Accessories        942
Furniture          941
Electronics        935
Name: count, dtype: int64


Output 2:
Category
Office Supplies    0.252718
Accessories        0.249801
Furniture          0.249536
Electronics        0.247945
Name: proportion, dtype: float64


##### Question 16
What is the difference between the two outputs?
- **Answer:** The first output shows the count of each catagory, while the second output displays the percentage each catagory makes up.

##### Question 17
What is our most common product category?
- **Answer:** Office supplies are the most common catagory.

##### Excercise 00
Using what you've learned, create a code cell direclty below this one to answer these questions:

- What is the date range of our sales data? (Hint: try `min()` and `max()` functions)
- How many unique products do we sell?
- What is our most common payment method?

In [24]:
print(f'The range of the sales data is from {(sales_df["Date"].max())} to {(sales_df["Date"].min())}')

The range of the sales data is from 2023-12-31 to 2023-01-01


In [26]:
print(sales_df['Product'].unique())

['Notebook' 'Laptop' 'Tablet' 'USB Drive' 'Pen Set' 'Laptop Bag' 'Lamp'
 'Smartwatch' 'Monitor Stand' 'Desk Organizer' 'Smartphone' 'Stapler'
 'Desk' 'Mouse' 'Keyboard' 'Office Chair' 'Bookshelf' 'Paper Clips'
 'Headphones' 'Filing Cabinet']


### Part 2 - Data Selection and Filtering

#### 02.01 - Column Selection
There are multiple ways to select columns in pandas. Let's explore them:

In [27]:
# Select single column
prices = sales_df['Price']

# Select multiple columns
product_info = sales_df[['Product', 'Price', 'Units']]

##### Question 17
Compare and contrast how we select multiple columns vs a single column.
- **Answer:** We select multuple columns by providing a list of column titles, rather than one column.

##### Excercise 01
Using what you've learned, create a code cell direclty below this one to select the 'Region' and 'Category' columns and store them in a variable called 'location_data'.

In [29]:
sales_df[['Region', 'Category']]

Region         Category
0     South  Office Supplies
1     North      Electronics
2      East      Electronics
3      East      Accessories
4      East  Office Supplies
...     ...              ...
3766  South      Electronics
3767  North        Furniture
3768   East      Electronics
3769  North        Furniture
3770  North  Office Supplies

[3771 rows x 2 columns]

#### 02.01 - Filtering Data
Let's learn how to filter our data based on conditions:

In [30]:
expensive_items = sales_df[sales_df['Price'] > 500]

north_sales = sales_df[sales_df['Region'] == 'North']

print("Expensive items:")
print(expensive_items)
print("\nNorth region sales:")
print(north_sales)

Expensive items:
            Date     Product     Category    Price  Units    Total  Discount  \
1     2023-01-01      Laptop  Electronics  1943.35      4  7773.40      0.00   
9     2023-01-01  Smartwatch  Electronics   893.71      3  2681.13      0.00   
10    2023-01-01      Laptop  Electronics   802.12      3  2093.53      0.13   
13    2023-01-02  Smartphone  Electronics   687.98      3  2063.94      0.00   
16    2023-01-02  Smartwatch  Electronics   507.91      4  2031.64      0.00   
...          ...         ...          ...      ...    ...      ...       ...   
3760  2023-12-31  Smartphone  Electronics  1766.20      4  7064.80      0.00   
3763  2023-12-31      Tablet  Electronics  1424.20      5  7121.00      0.00   
3764  2023-12-31  Smartphone  Electronics   857.34      5  4286.70      0.00   
3766  2023-12-31  Smartwatch  Electronics  1730.72      2  3461.44      0.00   
3768  2023-12-31      Laptop  Electronics   560.79      2  1121.58      0.00   

     Region          S

##### Question 18
What is `sales_df['Price'] > 500` accomplishing in our code? How might we use this in data science workflows?
- **Answer:** This will only display the items whose price exceeds 500. This can be used, for example, to isolate only very large and significant numbers in some kind of scientific study.

##### Question 19
What logic operator would we use to combine multiple conditonals when filtering columns?
- **Answer:** We would use the and operator to combine conditionals.

##### Excercise 02
Using what you've learned, create a code cell direclty below this one to create the following filters:
- The number of units sold was greater than 5
- The product category is 'Electronics'

In [35]:
sales_df[
    (sales_df['Category']=="Electronics") & (sales_df["Units"]>5)
    ]

Empty DataFrame
Columns: [Date, Product, Category, Price, Units, Total, Discount, Region, Sales_Rep, Payment_Method, Customer_Segment]
Index: []

In [41]:
sales_df=sales_df[[sales_df["Region"]=="North"]]
sales_df=sales_df

SyntaxError: '[' was never closed (3348023200.py, line 1)

#### Part 3 - Final Summary
Use your experience completing this lab, and the code block below to answer the following summary questions

##### Question 20
In your own words, explain what pandas is and why it's useful for data analysis:
- **Answer:** pandas is a useful tool for manipulating and sorting data. It allows us to get information from our data set that would be tedious to do manually -such as the mean of a specific column or filtering with specific parameters- automaticaly.

##### Question 21
How many unique products are in the dataset?
- **Answer:** There 20 unique products in the dataset

##### Question 22
What was the most popular product category in the `North` region? 
- **Answer:**